### ISSUE TITLE
Implement Feature Engineering Pipeline for Steel Casting Defect Prediction

### ISSUE THREAD

**Issue by dhar174** [dhar174/steel_defect_demo#24]
# Feature Engineering Pipeline Implementation

## Overview
Implement a comprehensive feature engineering pipeline that extracts meaningful patterns from time series sensor data for steel casting defect prediction. This pipeline will transform raw sensor readings into engineered features that capture statistical, stability, duration, interaction, and temporal patterns.

## Objectives
- Create a robust `CastingFeatureEngineer` class that processes time series sensor data
- Extract 80+ meaningful features from 5 sensor channels (temperature, pressure, flow_rate, vibration, power_consumption)
- Implement vectorized operations for efficient processing of large datasets
- Ensure features are properly scaled and normalized for machine learning models
- Provide comprehensive logging and error handling

### File Structure
```
src/features/
├── __init__.py
├── feature_engineer.py          # Main implementation
├── feature_utils.py            # Utility functions
└── feature_validation.py       # Data validation helpers
```

#### Class: `CastingFeatureEngineer`

**Dependencies:**
```python
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler, RobustScaler
from scipy import stats
from scipy.signal import find_peaks
import logging
from typing import Dict, List, Tuple, Optional, Union
```

**Class Structure:**
```python
class CastingFeatureEngineer:
    def __init__(self, 
                 sensor_columns: List[str] = None,
                 scaling_method: str = 'standard',
                 percentiles: List[float] = [10, 25, 75, 90],
                 spike_threshold: float = 2.0,
                 extreme_percentiles: Tuple[float, float] = (5, 95)):
        """
        Initialize the feature engineering pipeline
        
        Args:
            sensor_columns: List of sensor column names
            scaling_method: 'standard', 'robust', or 'none'
            percentiles: Percentiles to calculate for statistical features
            spike_threshold: Standard deviation threshold for spike detection
            extreme_percentiles: Lower and upper percentiles for extreme detection
        """
```

#### Required Methods:

**1. Statistical Features (35 features total)**
```python
def extract_statistical_features(self, df: pd.DataFrame) -> pd.DataFrame:
    """
    Extract statistical features for each sensor
    Features per sensor (7 × 5 sensors = 35 features):
    - mean, std, min, max, median
    - percentiles (10th, 25th, 75th, 90th)
    
    Returns:
        DataFrame with statistical features
    """
```

**2. Stability Features (20 features total)**
```python
def extract_stability_features(self, df: pd.DataFrame) -> pd.DataFrame:
    """
    Extract stability and variability features
    Features per sensor (4 × 5 sensors = 20 features):
    - spike_count: Number of values exceeding spike_threshold standard deviations
    - excursion_frequency: Rate of crossing mean ± 1 std
    - coefficient_of_variation: std/mean ratio
    - range_ratio: (max - min) / mean
    
    Returns:
        DataFrame with stability features
    """
```

**3. Duration Features (15 features total)**
```python
def extract_duration_features(self, df: pd.DataFrame) -> pd.DataFrame:
    """
    Extract duration-based features
    Features per sensor (3 × 5 sensors = 15 features):
    - time_at_extremes: Percentage of time spent in extreme percentiles
    - threshold_crossings: Number of mean crossings
    - consecutive_extremes: Maximum consecutive points in extreme regions
    
    Returns:
        DataFrame with duration features
    """
```

**4. Interaction Features (10 features total)**
```python
def extract_interaction_features(self, df: pd.DataFrame) -> pd.DataFrame:
    """
    Extract cross-sensor interaction features
    Features (10 total):
    - temp_pressure_corr: Correlation between temperature and pressure
    - temp_flow_ratio: Mean temperature / mean flow_rate
    - pressure_vibration_diff: Mean pressure - mean vibration
    - flow_power_corr: Correlation between flow_rate and power_consumption
    - temp_power_product: Mean temperature × mean power_consumption
    - pressure_flow_corr: Correlation between pressure and flow_rate
    - vibration_power_ratio: Mean vibration / mean power_consumption
    - temp_vibration_corr: Correlation between temperature and vibration
    - overall_sensor_corr: Mean correlation across all sensor pairs
    - sensor_variance_ratio: Max sensor variance / min sensor variance
    
    Returns:
        DataFrame with interaction features
    """
```

**5. Temporal Features (10 features total)**
```python
def extract_temporal_features(self, df: pd.DataFrame) -> pd.DataFrame:
    """
    Extract temporal pattern features
    Features per sensor (2 × 5 sensors = 10 features):
    - linear_trend: Slope of linear regression line
    - mean_gradient: Average absolute difference between consecutive points
    
    Returns:
        DataFrame with temporal features
    """
```

**6. Main Processing Method**
```python
def engineer_features(self, 
                     data: Union[pd.DataFrame, Dict],
                     cast_id: Optional[str] = None) -> pd.DataFrame:
    """
    Main method to extract all features from a casting sequence
    
    Args:
        data: DataFrame with sensor readings or dict of DataFrames
        cast_id: Optional casting ID for tracking
        
    Returns:
        DataFrame with all engineered features (80+ columns)
    """
```

**7. Batch Processing Method**
```python
def engineer_features_batch(self, 
                           data_dict: Dict[str, pd.DataFrame],
                           n_jobs: int = -1) -> pd.DataFrame:
    """
    Process multiple casting sequences in parallel
    
    Args:
        data_dict: Dictionary mapping cast_ids to DataFrames
        n_jobs: Number of parallel jobs
        
    Returns:
        DataFrame with features for all casts
    """
```

### Feature Validation: `src/features/feature_validation.py`

```python
class FeatureValidator:
    """Validate input data and engineered features"""
    
    def validate_input_data(self, df: pd.DataFrame) -> bool:
        """Validate input sensor data format and content"""
        
    def validate_features(self, features_df: pd.DataFrame) -> Dict[str, bool]:
        """Validate engineered features for completeness and quality"""
        
    def check_feature_correlations(self, features_df: pd.DataFrame) -> pd.DataFrame:
        """Check for highly correlated features that may need removal"""
```

### Utility Functions: `src/features/feature_utils.py`

```python
def calculate_percentiles(series: pd.Series, percentiles: List[float]) -> Dict[str, float]:
    """Calculate multiple percentiles efficiently"""

def detect_spikes(series: pd.Series, threshold: float = 2.0) -> np.ndarray:
    """Detect spikes using z-score method"""

def count_threshold_crossings(series: pd.Series, threshold: float) -> int:
    """Count number of times series crosses a threshold"""

def calculate_trend_slope(series: pd.Series) -> float:
    """Calculate linear trend slope using least squares"""

def safe_correlation(x: pd.Series, y: pd.Series) -> float:
    """Calculate correlation with NaN handling"""
```

### Feature Categories and Counts:
1. **Statistical Features**: 35 (7 stats × 5 sensors)
2. **Stability Features**: 20 (4 metrics × 5 sensors)
3. **Duration Features**: 15 (3 metrics × 5 sensors)
4. **Interaction Features**: 10 (cross-sensor relationships)
5. **Temporal Features**: 10 (2 metrics × 5 sensors)

**Total Features**: 90 engineered features

### Feature Naming Convention:
```
Statistical: {sensor}_mean, {sensor}_std, {sensor}_min, {sensor}_max, {sensor}_median, {sensor}_p10, {sensor}_p25, {sensor}_p75, {sensor}_p90
Stability: {sensor}_spike_count, {sensor}_excursion_freq, {sensor}_cv, {sensor}_range_ratio
Duration: {sensor}_time_extremes, {sensor}_threshold_crossings, {sensor}_consec_extremes
Interaction: temp_pressure_corr, temp_flow_ratio, pressure_vibration_diff, etc.
Temporal: {sensor}_trend, {sensor}_gradient
```

### Performance Requirements:
- Process 1,200 casting sequences in < 2 minutes
- Memory usage < 1GB for full dataset processing
- Support parallel processing for batch operations
- Efficient vectorized operations using NumPy/Pandas

### Error Handling:
- Graceful handling of missing sensor data
- NaN/infinity value detection and replacement
- Input validation with informative error messages
- Logging of processing steps and warnings

### Data Quality Checks:
- Validate sensor data completeness (minimum 50 data points per sequence)
- Check for constant/flatline sensor readings
- Detect and flag potential sensor malfunctions
- Ensure feature values are within reasonable ranges

### Unit Tests (`tests/test_feature_engineering.py`):

```python
class TestCastingFeatureEngineer:
    def test_statistical_features_calculation(self):
        """Test statistical feature extraction accuracy"""
        
    def test_stability_features_edge_cases(self):
        """Test stability features with edge cases (constant values, spikes)"""
        
    def test_duration_features_boundary_conditions(self):
        """Test duration features with various data patterns"""
        
    def test_interaction_features_correlation_calculation(self):
        """Test cross-sensor interaction features"""
        
    def test_temporal_features_trend_detection(self):
        """Test temporal pattern extraction"""
        
    def test_batch_processing_consistency(self):
        """Test batch processing produces same results as individual processing"""
        
    def test_missing_data_handling(self):
        """Test graceful handling of missing sensor data"""
        
    def test_feature_scaling_options(self):
        """Test different scaling methods"""
        
    def test_performance_benchmarks(self):
        """Test processing speed meets requirements"""
```

### Integration Tests:
- Test with actual synthetic data from Phase 1
- Verify feature engineering integrates with model training pipeline
- Test memory usage with full dataset
- Validate feature importance rankings make sense

### Configuration File: `config/feature_engineering.yaml`
```yaml
feature_engineering:
  sensor_columns:
    - temperature
    - pressure
    - flow_rate
    - vibration
    - power_consumption
  
  statistical_features:
    percentiles: [10, 25, 75, 90]
    
  stability_features:
    spike_threshold: 2.0
    excursion_std_multiplier: 1.0
    
  duration_features:
    extreme_percentiles: [5, 95]
    
  scaling:
    method: "standard"  # standard, robust, none
    
  validation:
    min_data_points: 50
    max_missing_ratio: 0.1
    
  performance:
    batch_size: 100
    n_jobs: -1
```

### Basic Usage:
```python
from src.features.feature_engineer import CastingFeatureEngineer

# Initialize engineer
engineer = CastingFeatureEngineer()

# Load casting data
casting_data = pd.read_csv('data/synthetic/casting_sequence_001.csv')

# Extract features
features = engineer.engineer_features(casting_data, cast_id='001')
print(f"Extracted {len(features.columns)} features")
```

### Batch Processing:
```python
# Load multiple castings
data_dict = {}
for cast_id in cast_ids:
    data_dict[cast_id] = pd.read_csv(f'data/synthetic/casting_sequence_{cast_id}.csv')

# Process all at once
all_features = engineer.engineer_features_batch(data_dict)
```

## Success Criteria
- [ ] `CastingFeatureEngineer` class successfully extracts 90+ features
- [ ] All feature extraction methods handle edge cases gracefully
- [ ] Processing speed meets performance requirements (< 2 minutes for 1,200 sequences)
- [ ] Unit tests achieve > 90% code coverage
- [ ] Integration tests pass with synthetic data from Phase 1
- [ ] Feature validation detects and handles data quality issues
- [ ] Batch processing produces consistent results
- [ ] Memory usage stays within specified limits
- [ ] Configuration system allows easy parameter tuning
- [ ] Documentation includes clear usage examples

## Dependencies
```
pandas>=1.5.0
numpy>=1.23.0
scikit-learn>=1.1.0
scipy>=1.9.0
joblib>=1.2.0
pyyaml>=6.0
pytest>=7.2.0
```

## Related Issues
- Depends on: Phase 1 Synthetic Data Generation
- Blocks: #X (Baseline Model Implementation)
- Blocks: #Y (Model Training Pipeline)

## Assignee Suggestions
Assign to team members with experience in:
- Time series feature engineering
- Pandas/NumPy optimization
- Statistical signal processing
- Machine learning pipeline development
